# Capstone Project - The Battle of Neighborhoods

## Business Problem Section

#### Background

London is the largest city and the capital of United Kingdom. It is considered as one of the most important global cities. Its estimated mid-2018 municipal population (corresponding to Greater London) was 8,908,081, the third most populous of any city in Europe and accounts for 13.4% of the UK population. The city covers a total area of 607 sq. miles. Clearly, London is a city with high population density. Real estate investment in such cities thus becomes a big investment. And before making such an investment, one should have a proper investment strategy.

#### Business Problem

In a country with such a huge area and high real estate prices, it becomes very tough to find a property with its appropriate value, keeping in mind all the amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores. As a result, the business problem we are currently posing is: How could we recommend profitable venues to support homebuyers according to amenities and essential facilities surrounding such venues?

## Data Section

Data on London properties and the relative price paid data were extracted from the HM Land Registry (http://landregistry.data.gov.uk/).
The following fields comprise the address data included in Price Paid Data: Postcode; PAON Primary Addressable Object Name, typically the house number or name; SAON Secondary Addressable Object Name. If there is a sub-building, for example, the building is divided into flats, there will be a SAON; Street; Locality; Town/City; District; County.

To explore and target recommended locations across different venues according to the presence of amenities and essential facilities, we will access data through FourSquare API interface and arrange them as a dataframe for visualization.

By merging data on London properties and the relative price paid data from the HM Land Registry and data on amenities and essential facilities surrounding such properties from FourSquare API interface, we will be able to recommend profitable real estate investments.

## Methodology

This section will describe the main componeants of our anaysis and prediction system. It consisits of the following sections:
1. Data Collection
2. Exploring and understanding the data.
3. Data preparation and processing
4. Data Modeling

#### 1. Data Collection

We'll start with first importing the required libraries.

In [2]:
import numpy as np
import pandas as pd
import datetime as dt # Datetime
import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

!conda install -c conda-forge folium=0.5.0 --yes
import folium #import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

Now we will downlod data from the HM Land Registry

In [3]:
df_ppd = pd.read_csv("http://prod2.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com/pp-2018.csv")

#### 2. Exploring and understanding the data.

Before using the data, it is important to understand it.
Let's check the first five rows to see what our data contains.

In [4]:
df_ppd.head()

{79A74E21-D11E-1289-E053-6B04A8C01627}  770000  2018-09-25 00:00  SK7 1AR  \
0  {79A74E21-D11F-1289-E053-6B04A8C01627}  253500  2018-09-24 00:00   M6 8GQ   
1  {79A74E21-D120-1289-E053-6B04A8C01627}  231950  2018-09-28 00:00  WA3 2UE   
2  {79A74E21-D121-1289-E053-6B04A8C01627}  112500  2018-08-29 00:00  OL6 6RJ   
3  {79A74E21-D122-1289-E053-6B04A8C01627}  184995  2018-06-15 00:00  M46 0TW   
4  {79A74E21-D123-1289-E053-6B04A8C01627}  214995  2018-09-28 00:00  M28 3XS   

   D  N  F    5 Unnamed: 8          OAK MEADOW  BRAMHALL          STOCKPORT  \
0  D  N  F    1        NaN      RIVINGTON ROAD       NaN            SALFORD   
1  D  Y  F   35        NaN     STONEACRE CLOSE    LOWTON         WARRINGTON   
2  S  N  F  102        NaN    THORNFIELD GROVE       NaN  ASHTON-UNDER-LYNE   
3  S  Y  F   37        NaN  THREADNEEDLE PLACE  ATHERTON         MANCHESTER   
4  D  Y  L    9        NaN      MARPLE GARDENS   WORSLEY         MANCHESTER   

  STOCKPORT.1  GREATER MANCHESTER  A A.1  
0     SALFORD  GREATER MANCHESTER  A   A  
1       WIGAN  GREATER MANCHESTER  A   A  
2    TAMESIDE  GREATER MANCHESTER  A   A  
3       WIGAN  GREATER MANCHESTER  A   A  
4     SALFORD  GREATER MANCHESTER  A   A

In [5]:
df_ppd.shape

(1029929, 16)

Our data has 16 columns and more than 1,000,000 rows. We shall now prepare and preprocess data acoording to our requirement.

#### 3. Preparing and Processing the data

At this stage we convert our data from intial raw form into another preprocessed form inorder to prepare the data for further analysis. Therefore we perform the following steps:
- Rename the columns
- Format the date column and sort the data by date of sale
- Make a list of street names in London
- Calculate the street-wise average price of the property
- Read the street-wise coordinates into a data frame, eliminating re-occuring word London from the indivisual names
- Join the data to find the coordinates that fits into our client's budget
- Plot recommended locations on London map along with cureent market prices

Assign meaningful names to the columns

In [64]:
df_ppd.columns=['TUID', 'Price', 'Date_Transfer', 'Postcode', 'Prop_Type', 'Old_New', 'Duration', 'PAON', \
                  'SAON', 'Street', 'Locality', 'Town_City', 'District', 'County', 'PPD_Cat_Type', 'Record_Status']

Next step is to format the date column

In [7]:
df_ppd['Date_Transfer']=df_ppd['Date_Transfer'].apply(pd.to_datetime)

In [8]:
# Delete all obsolete transactions which were done before 2018
df_ppd.drop(df_ppd[df_ppd.Date_Transfer.dt.year < 2018].index, inplace=True)

# Sort by Date of Sale
df_ppd.sort_values(by=['Date_Transfer'],ascending=[False],inplace=True)

Now, lets list down the streets in London

In [9]:
df_ppd_london = df_ppd.query("Town_City == 'LONDON'")
streets = df_ppd_london['Street'].unique().tolist()

In [10]:
streets

['BRIDGE LANE',
 'HERMITAGE ROAD',
 'CAMLEY STREET',
 'ST JOSEPHS STREET',
 'SHEPHERDESS WALK',
 'DERWENT ROAD',
 'VICTORIA STREET',
 'BROSTER GARDENS',
 'SYLVAN HILL',
 'THREE OAK LANE',
 'GLASSHOUSE GARDENS',
 'QUICKS ROAD',
 'ST GEORGES SQUARE',
 'HYDE ESTATE ROAD',
 'RITHERDON ROAD',
 'TRINITY CRESCENT',
 'APPLE YARD',
 'LANCASTER GATE',
 'HANDYSIDE STREET',
 'MICHLEHAM DOWN',
 'CHINGFORD MOUNT ROAD',
 'GRAND DRIVE',
 'EXCHANGE GARDENS',
 'BRUSHFIELD STREET',
 'WENTWORTH STREET',
 'BUCKINGHAM PALACE ROAD',
 'CLIFFORD STREET',
 'HOWARD ROAD',
 'BLACKFRIARS ROAD',
 'HIGHWOOD HILL',
 "ST JAMES'S PLACE",
 'FLORAL STREET',
 'RYMILL STREET',
 'COLLENDALE ROAD',
 'TURNHAM GREEN TERRACE',
 'PALACE COURT',
 'HONOUR LEA AVENUE',
 'ARCHWAY ROAD',
 'HARTFIELD ROAD',
 'GALLOWAY ROAD',
 'WARWICK WAY',
 'OXFORD STREET',
 'UPPER TERRACE',
 'WOOLF CLOSE',
 'FULHAM HIGH STREET',
 'ABBOTS MANOR',
 'LOWER BOSTON ROAD',
 'PRINCE ALBERT ROAD',
 'UPPER RICHMOND ROAD',
 'NORTH BIRKBECK ROAD',
 'CLIFFORD R

We now calculate the street-wise average price of the property.

In [11]:
df_grp_price = df_ppd_london.groupby(['Street'])['Price'].mean().reset_index()

# Giving column names
df_grp_price.columns = ['Street', 'Avg_Price']

In [12]:
#Input your Budget's Upper Limit and Lower Limit - Find the locations df_grp_price which fits your budget
df_affordable = df_grp_price.query("(Avg_Price >= 2200000) & (Avg_Price <= 2500000)")
df_affordable

Street     Avg_Price
196            ALBION SQUARE  2.450000e+06
390              ANHALT ROAD  2.435000e+06
405          ANSDELL TERRACE  2.250000e+06
422          APPLEGARTH ROAD  2.400000e+06
855          BARONSMEAD ROAD  2.375000e+06
981           BEAUCLERC ROAD  2.480000e+06
1102         BELVEDERE DRIVE  2.340000e+06
1215       BICKENHALL STREET  2.208500e+06
1253       BIRCHLANDS AVENUE  2.217000e+06
1553          BRAMPTON GROVE  2.456875e+06
1632       BRIARDALE GARDENS  2.397132e+06
1797                BROOKWAY  2.400000e+06
1914            BURBAGE ROAD  2.445000e+06
1980               BURY WALK  2.492500e+06
2068         CALLCOTT STREET  2.375000e+06
2129       CAMPDEN HILL ROAD  2.379653e+06
2136            CAMPION ROAD  2.461000e+06
2158           CANNING PLACE  2.425000e+06
2225           CARLISLE ROAD  2.200000e+06
2230         CARLTON GARDENS  2.483500e+06
2242           CARLYLE COURT  2.300000e+06
2406          CHALCOT SQUARE  2.286679e+06
2484            CHARLES LANE  2.414000e+06
2562        CHELSEA CRESCENT  2.495000e+06
2606     CHESTER CLOSE NORTH  2.450000e+06
2638            CHEYNE COURT  2.250000e+06
2641              CHEYNE ROW  2.410000e+06
2686           CHISWICK MALL  2.287500e+06
2761               CITY ROAD  2.468340e+06
2808        CLARENDON STREET  2.250000e+06
...                      ...           ...
10935   RUSSELL GARDENS MEWS  2.300000e+06
11183         SETTLES STREET  2.487500e+06
11258         SHELDON AVENUE  2.349542e+06
11492          SOUTH END ROW  2.470000e+06
11569    SOUTHWOOD LAWN ROAD  2.350000e+06
11572         SOVEREIGN PARK  2.500000e+06
11789  ST MARGARETS CRESCENT  2.216500e+06
11825       ST OSWALDS PLACE  2.250000e+06
11845       ST PETERS SQUARE  2.468730e+06
11882       STAFFORD TERRACE  2.355000e+06
12222       SUTHERLAND PLACE  2.456000e+06
12284          SYDNEY STREET  2.240833e+06
12426            THAMES BANK  2.400000e+06
12488            THE HEXAGON  2.335000e+06
12767        TREDEGAR SQUARE  2.436666e+06
12829         TRINITY STREET  2.317500e+06
12979   UPPER HAMPSTEAD WALK  2.500000e+06
13241        WALPOLE GARDENS  2.303500e+06
13243         WALPOLE STREET  2.242500e+06
13317         WARWICK SQUARE  2.432273e+06
13416            WELBECK WAY  2.267000e+06
13428      WELLESLEY TERRACE  2.410000e+06
13439      WELLINGTON STREET  2.293155e+06
13571     WESTMORELAND PLACE  2.300000e+06
13678       WHITFIELD STREET  2.451000e+06
13723         WILFRED STREET  2.410538e+06
13749     WILLOW BRIDGE ROAD  2.425000e+06
13769          WILSON STREET  2.257500e+06
13797        WINCHENDON ROAD  2.350000e+06
13834           WINGATE ROAD  2.206400e+06

[162 rows x 2 columns]

We have 162 street locations that fits into our clint's budget

In [13]:
import pandas as pd
import numpy as np
import datetime as DT
import hmac
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# import k-means from clustering stage
from sklearn.cluster import KMeans

In [14]:
geolocator = Nominatim(user_agent="ldn_explorer")
df_affordable['city_coord'] = df_affordable['Street'].apply(geolocator.geocode).apply(lambda x: (x.latitude, x.longitude))
df_affordable[['Latitude', 'Longitude']] = df_affordable['city_coord'].apply(pd.Series)
df = df_affordable.drop(columns=['city_coord'])
df

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/frame.py:3391: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


Street     Avg_Price   Latitude   Longitude
196            ALBION SQUARE  2.450000e+06 -41.273758  173.289393
390              ANHALT ROAD  2.435000e+06  51.480316   -0.166801
405          ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103
422          APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670
855          BARONSMEAD ROAD  2.375000e+06  51.477315   -0.239457
981           BEAUCLERC ROAD  2.480000e+06  30.211452  -81.617981
1102         BELVEDERE DRIVE  2.340000e+06  44.707562  -63.545599
1215       BICKENHALL STREET  2.208500e+06  51.521201   -0.158908
1253       BIRCHLANDS AVENUE  2.217000e+06  51.448394   -0.160468
1553          BRAMPTON GROVE  2.456875e+06  51.589961   -0.318525
1632       BRIARDALE GARDENS  2.397132e+06  51.560175   -0.195431
1797                BROOKWAY  2.400000e+06  45.432185 -122.802812
1914            BURBAGE ROAD  2.445000e+06  52.538507   -1.353674
1980               BURY WALK  2.492500e+06  52.145529   -0.423593
2068         CALLCOTT STREET  2.375000e+06  51.508350   -0.198328
2129       CAMPDEN HILL ROAD  2.379653e+06  51.506352   -0.198853
2136            CAMPION ROAD  2.461000e+06  43.623321  -70.236225
2158           CANNING PLACE  2.425000e+06  51.499494   -0.185481
2225           CARLISLE ROAD  2.200000e+06  43.399438  -79.973215
2230         CARLTON GARDENS  2.483500e+06 -37.801943  144.971970
2242           CARLYLE COURT  2.300000e+06  32.972701  -97.173392
2406          CHALCOT SQUARE  2.286679e+06  51.541088   -0.155928
2484            CHARLES LANE  2.414000e+06  40.257626  -75.093891
2562        CHELSEA CRESCENT  2.495000e+06  34.522443  -85.443891
2606     CHESTER CLOSE NORTH  2.450000e+06  51.529205   -0.145081
2638            CHEYNE COURT  2.250000e+06  52.035905    0.727612
2641              CHEYNE ROW  2.410000e+06  51.483717   -0.169603
2686           CHISWICK MALL  2.287500e+06  51.487994   -0.246605
2761               CITY ROAD  2.468340e+06  51.529697   -0.097763
2808        CLARENDON STREET  2.250000e+06  42.284961  -71.345534
...                      ...           ...        ...         ...
10935   RUSSELL GARDENS MEWS  2.300000e+06  51.499918   -0.211944
11183         SETTLES STREET  2.487500e+06  32.239424 -101.457123
11258         SHELDON AVENUE  2.349542e+06  36.981254 -122.040507
11492          SOUTH END ROW  2.470000e+06  51.498746   -0.189079
11569    SOUTHWOOD LAWN ROAD  2.350000e+06  51.574627   -0.146238
11572         SOVEREIGN PARK  2.500000e+06 -27.493713  153.210143
11789  ST MARGARETS CRESCENT  2.216500e+06  17.097514  -88.616115
11825       ST OSWALDS PLACE  2.250000e+06  51.487207   -0.118534
11845       ST PETERS SQUARE  2.468730e+06  41.902235   12.457357
11882       STAFFORD TERRACE  2.355000e+06  58.014299   -3.855648
12222       SUTHERLAND PLACE  2.456000e+06  44.310685  -78.309861
12284          SYDNEY STREET  2.240833e+06 -29.823972   30.919248
12426            THAMES BANK  2.400000e+06  52.085669   -0.243276
12488            THE HEXAGON  2.335000e+06  51.453882   -0.977834
12767        TREDEGAR SQUARE  2.436666e+06  51.527093   -0.032184
12829         TRINITY STREET  2.317500e+06  50.338316   -4.795430
12979   UPPER HAMPSTEAD WALK  2.500000e+06  51.558467   -0.177453
13241        WALPOLE GARDENS  2.303500e+06  51.439743   -0.341263
13243         WALPOLE STREET  2.242500e+06 -27.872584  151.271777
13317         WARWICK SQUARE  2.432273e+06  33.738563 -117.846506
13416            WELBECK WAY  2.267000e+06  52.558005   -0.262309
13428      WELLESLEY TERRACE  2.410000e+06  54.974313   -1.640202
13439      WELLINGTON STREET  2.293155e+06  45.423449  -75.698058
13571     WESTMORELAND PLACE  2.300000e+06  40.740224 -111.846323
13678       WHITFIELD STREET  2.451000e+06  51.521833   -0.136734
13723         WILFRED STREET  2.410538e+06  51.442178    0.372743
13749     WILLOW BRIDGE ROAD  2.425000e+06  53.640874   -1.200462
13769          WILSON STREET  2.257500e+06  45.700486 -121.521875
13797        WINCHENDON ROAD  2.350000e+06  42.645331  -71.954

In [15]:

address = 'London, UK'

geolocator = Nominatim(user_agent="ldn_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of London City are 51.5073219, -0.1276474.


In [16]:
map_london = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, price, street in zip(df['Latitude'], df['Longitude'], df['Avg_Price'], df['Street']):
    label = '{}, {}'.format(street, price)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_london)  
    
map_london

In [17]:
#Define Foursquare Credentials and Version

CLIENT_ID = 'AG4W05YY1I5QC5TFIGI3AXDTDMTAEDEELHNW2SFQJAA5EELF' # your Foursquare ID
CLIENT_SECRET = 'CO1HYE001KSTDVZ2TG5GSLN0M2YGWATVUBQSCHGGNU5XL1R4' 
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: AG4W05YY1I5QC5TFIGI3AXDTDMTAEDEELHNW2SFQJAA5EELF
CLIENT_SECRET:CO1HYE001KSTDVZ2TG5GSLN0M2YGWATVUBQSCHGGNU5XL1R4


We have located the affordable coordinated into the map. We can now proceed to the Modeling phase. We will analyze the neighborhoods to recommend real estates where hom buyers can make their investments. We will then recommend them profitable venues according to amenities and essential facilities in the surrounding area.

#### 4. Modeling

We will use the K-Means technique as it is the fastest and efficient in terms of computational cost, is highly flexible to account for mutations in real estate market in London.

In [18]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Street', 
                  'Street Latitude', 
                  'Street Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [19]:
# Run the above function on each location and create a new dataframe called location_venues and display it.
location_venues = getNearbyVenues(names=df['Street'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

ALBION SQUARE
ANHALT ROAD
ANSDELL TERRACE
APPLEGARTH ROAD
BARONSMEAD ROAD
BEAUCLERC ROAD
BELVEDERE DRIVE
BICKENHALL STREET
BIRCHLANDS AVENUE
BRAMPTON GROVE
BRIARDALE GARDENS
BROOKWAY
BURBAGE ROAD
BURY WALK
CALLCOTT STREET
CAMPDEN HILL ROAD
CAMPION ROAD
CANNING PLACE
CARLISLE ROAD
CARLTON GARDENS
CARLYLE COURT
CHALCOT SQUARE
CHARLES LANE
CHELSEA CRESCENT
CHESTER CLOSE NORTH
CHEYNE COURT
CHEYNE ROW
CHISWICK MALL
CITY ROAD
CLARENDON STREET
CLONCURRY STREET
COLBECK MEWS
COLLEGE CRESCENT
CORNWALL TERRACE MEWS
COURT LANE GARDENS
CRESCENT GROVE
DALEBURY ROAD
DEWHURST ROAD
DORIA ROAD
DOWNSHIRE HILL
DUCHESS WALK
ECCLESTON SQUARE MEWS
EGBERT STREET
EGERTON PLACE
ELM PARK ROAD
FLORAL STREET
FRANK DIXON WAY
FULTON MEWS
GERARD ROAD
GERRARD ROAD
GIRDLERS ROAD
GLOUCESTER CRESCENT
GORDON PLACE
GRAFTON SQUARE
GRAHAM TERRACE
HARMAN DRIVE
HARRIS STREET
HAVANNAH STREET
HAZLEWELL ROAD
HEREFORD MEWS
HERONDALE AVENUE
HIGHGATE HIGH STREET
HIGHWOOD HILL
HILLGATE PLACE
HOLLYCROFT AVENUE
HOLLYWOOD MEWS
HONEYWELL

In [20]:
location_venues

Street  Street Latitude  Street Longitude  \
0      ALBION SQUARE       -41.273758        173.289393   
1      ALBION SQUARE       -41.273758        173.289393   
2      ALBION SQUARE       -41.273758        173.289393   
3      ALBION SQUARE       -41.273758        173.289393   
4      ALBION SQUARE       -41.273758        173.289393   
5      ALBION SQUARE       -41.273758        173.289393   
6      ALBION SQUARE       -41.273758        173.289393   
7      ALBION SQUARE       -41.273758        173.289393   
8      ALBION SQUARE       -41.273758        173.289393   
9      ALBION SQUARE       -41.273758        173.289393   
10     ALBION SQUARE       -41.273758        173.289393   
11     ALBION SQUARE       -41.273758        173.289393   
12     ALBION SQUARE       -41.273758        173.289393   
13     ALBION SQUARE       -41.273758        173.289393   
14     ALBION SQUARE       -41.273758        173.289393   
15     ALBION SQUARE       -41.273758        173.289393   
16     ALBION SQUARE       -41.273758        173.289393   
17     ALBION SQUARE       -41.273758        173.289393   
18     ALBION SQUARE       -41.273758        173.289393   
19     ALBION SQUARE       -41.273758        173.289393   
20     ALBION SQUARE       -41.273758        173.289393   
21     ALBION SQUARE       -41.273758        173.289393   
22     ALBION SQUARE       -41.273758        173.289393   
23     ALBION SQUARE       -41.273758        173.289393   
24     ALBION SQUARE       -41.273758        173.289393   
25     ALBION SQUARE       -41.273758        173.289393   
26       ANHALT ROAD        51.480316         -0.166801   
27       ANHALT ROAD        51.480316         -0.166801   
28       ANHALT ROAD        51.480316         -0.166801   
29       ANHALT ROAD        51.480316         -0.166801   
...              ...              ...               ...   
4502  WILFRED STREET        51.442178          0.372743   
4503  WILFRED STREET        51.442178          0.372743   
4504  WILFRED STREET        51.442178          0.372743   
4505  WILFRED STREET        51.442178          0.372743   
4506  WILFRED STREET        51.442178          0.372743   
4507  WILFRED STREET        51.442178          0.372743   
4508  WILFRED STREET        51.442178          0.372743   
4509  WILFRED STREET        51.442178          0.372743   
4510  WILFRED STREET        51.442178          0.372743   
4511   WILSON STREET        45.700486       -121.521875   
4512   WILSON STREET        45.700486       -121.521875   
4513   WILSON STREET        45.700486       -121.521875   
4514   WILSON STREET        45.700486       -121.521875   
4515   WILSON STREET        45.700486       -121.521875   
4516   WILSON STREET        45.700486       -121.521875   
4517   WILSON STREET        45.700486       -121.521875   
4518   WILSON STREET        45.700486       -121.521875   
4519   WILSON STREET        45.700486       -121.521875   
4520   WILSON STREET        45.700486       -121.521875   
4521   WILSON STREET        45.700486       -121.521875   
4522   WILSON STREET        45.700486       -121.521875   
4523   WILSON STREET        45.700486       -121.521875   
4524   WILSON STREET        45.700486       -121.521875   
4525   WILSON STREET        45.700486       -121.521875   
4526   WILSON STREET        45.700486       -121.521875   
4527   WILSON STREET        45.700486       -121.521875   
4528    WINGATE ROAD       -29.922130         30.966309   
4529    WINGATE ROAD       -29.922130         30.966309   
4530    WINGATE ROAD       -29.922130         30.966309   
4531    WINGATE ROAD       -29.922130         30.966309   

                             Venue  Venue Latitude  Venue Longitude  \
0                   The Free House      -41.273340       173.287364   
1                  The Indian Cafe      -41.273308       173.286530   
2                  Queen's Gardens      -41.273671       173.291383   
3                            Urban      -41.274355       173.286317   


In [21]:
location_venues.groupby('Street').count()

Street Latitude  Street Longitude  Venue  \
Street                                                           
ALBION SQUARE                      26                26     26   
ANHALT ROAD                        15                15     15   
ANSDELL TERRACE                    45                45     45   
APPLEGARTH ROAD                     4                 4      4   
BARONSMEAD ROAD                    14                14     14   
BEAUCLERC ROAD                      4                 4      4   
BELVEDERE DRIVE                     1                 1      1   
BICKENHALL STREET                  61                61     61   
BIRCHLANDS AVENUE                  11                11     11   
BRAMPTON GROVE                      2                 2      2   
BRIARDALE GARDENS                   5                 5      5   
BURBAGE ROAD                        6                 6      6   
BURY WALK                          15                15     15   
CALLCOTT STREET                    49                49     49   
CAMPDEN HILL ROAD                  50                50     50   
CAMPION ROAD                        3                 3      3   
CANNING PLACE                      45                45     45   
CARLTON GARDENS                    65                65     65   
CARLYLE COURT                       2                 2      2   
CHALCOT SQUARE                     52                52     52   
CHARLES LANE                        4                 4      4   
CHESTER CLOSE NORTH                22                22     22   
CHEYNE COURT                       12                12     12   
CHEYNE ROW                         58                58     58   
CHISWICK MALL                       7                 7      7   
CITY ROAD                          27                27     27   
CLARENDON STREET                   29                29     29   
CLONCURRY STREET                   32                32     32   
COLBECK MEWS                      100               100    100   
COLLEGE CRESCENT                   16                16     16   
...                               ...               ...    ...   
RINGWOOD AVENUE                     4                 4      4   
RODERICK ROAD                       7                 7      7   
ROPEMAKERS FIELDS                  23                23     23   
ROYAL CRESCENT                     21                21     21   
RUSSELL GARDENS MEWS               25                25     25   
SETTLES STREET                      3                 3      3   
SHELDON AVENUE                      6                 6      6   
SOUTH END ROW                      47                47     47   
SOUTHWOOD LAWN ROAD                33                33     33   
SOVEREIGN PARK                      3                 3      3   
ST OSWALDS PLACE                   38                38     38   
ST PETERS SQUARE                   73                73     73   
STAFFORD TERRACE                    4                 4      4   
SUTHERLAND PLACE                    7                 7      7   
SYDNEY STREET                       3                 3      3   
THAMES BANK                         4                 4      4   
THE HEXAGON                        37                37     37   
TREDEGAR SQUARE                    27                27     27   
TRINITY STREET                     14                14     14   
UPPER HAMPSTEAD WALK               59                59     59   
WALPOLE GARDENS                     8                 8      8   
WARWICK SQUARE                      5                 5      5   
WELBECK WAY                         4                 4      4   
WELLESLEY TERRACE                  12                12     12   
WELLINGTON STREET                  85                85     85   
WESTMORELAND PLACE                 17                17     17   
WHITFIELD STREET                   84                84     84   
WILFRED STREET                     26                26     26

In [22]:
# get the List of Unique Categories
print('There are {} uniques categories.'.format(len(location_venues['Venue Category'].unique())))

There are 348 uniques categories.


In [23]:
# one hot encoding
venues_onehot = pd.get_dummies(location_venues[['Venue Category']], prefix="", prefix_sep="")

# add street column back to dataframe
venues_onehot['Street'] = location_venues['Street'] 

# move street column to the first column
fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])

#fixed_columns
venues_onehot = venues_onehot[fixed_columns]

venues_onehot.head()

Street  ATM  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  ALBION SQUARE    0                  0               0                  0   
1  ALBION SQUARE    0                  0               0                  0   
2  ALBION SQUARE    0                  0               0                  0   
3  ALBION SQUARE    0                  0               0                  0   
4  ALBION SQUARE    0                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Argentinian Restaurant  ...  Vietnamese Restaurant  Warehouse Store  \
0                       0  ...                      0                0   
1                       0  ...                      0                0   
2                       0  ...                      0                0   
3                       0  ...                      0                0   
4                       0  ...                      0                0   

   Waterfront  Weight Loss Center  Wine Bar  Wine Shop  Women's Store  \
0           0                   0         0          0              0   
1           0                   0         0          0              0   
2           0                   0         0          0              0   
3           0                   0         0          0              0   
4           0                   0         0          0              0   

   Yoga Studio  Zoo  Zoo Exhibit  
0            0    0            0  
1            0    0            0  
2            0    0            0  
3            0    0            0  
4            0    0            0  

[5 rows x 349 columns]

In [24]:
london_grouped = venues_onehot.groupby('Street').mean().reset_index()
london_grouped

Street       ATM  Accessories Store  Adult Boutique  \
0           ALBION SQUARE  0.000000           0.000000             0.0   
1             ANHALT ROAD  0.000000           0.000000             0.0   
2         ANSDELL TERRACE  0.000000           0.000000             0.0   
3         APPLEGARTH ROAD  0.000000           0.000000             0.0   
4         BARONSMEAD ROAD  0.000000           0.000000             0.0   
5          BEAUCLERC ROAD  0.000000           0.000000             0.0   
6         BELVEDERE DRIVE  0.000000           0.000000             0.0   
7       BICKENHALL STREET  0.000000           0.000000             0.0   
8       BIRCHLANDS AVENUE  0.000000           0.000000             0.0   
9          BRAMPTON GROVE  0.000000           0.000000             0.0   
10      BRIARDALE GARDENS  0.000000           0.000000             0.0   
11           BURBAGE ROAD  0.000000           0.000000             0.0   
12              BURY WALK  0.000000           0.000000             0.0   
13        CALLCOTT STREET  0.000000           0.000000             0.0   
14      CAMPDEN HILL ROAD  0.000000           0.000000             0.0   
15           CAMPION ROAD  0.000000           0.000000             0.0   
16          CANNING PLACE  0.000000           0.000000             0.0   
17        CARLTON GARDENS  0.000000           0.000000             0.0   
18          CARLYLE COURT  0.000000           0.000000             0.0   
19         CHALCOT SQUARE  0.000000           0.000000             0.0   
20           CHARLES LANE  0.000000           0.000000             0.0   
21    CHESTER CLOSE NORTH  0.000000           0.000000             0.0   
22           CHEYNE COURT  0.000000           0.000000             0.0   
23             CHEYNE ROW  0.000000           0.017241             0.0   
24          CHISWICK MALL  0.000000           0.000000             0.0   
25              CITY ROAD  0.000000           0.000000             0.0   
26       CLARENDON STREET  0.000000           0.000000             0.0   
27       CLONCURRY STREET  0.000000           0.000000             0.0   
28           COLBECK MEWS  0.000000           0.000000             0.0   
29       COLLEGE CRESCENT  0.000000           0.000000             0.0   
..                    ...       ...                ...             ...   
119       RINGWOOD AVENUE  0.000000           0.000000             0.0   
120         RODERICK ROAD  0.000000           0.000000             0.0   
121     ROPEMAKERS FIELDS  0.000000           0.000000             0.0   
122        ROYAL CRESCENT  0.000000           0.000000             0.0   
123  RUSSELL GARDENS MEWS  0.000000           0.000000             0.0   
124        SETTLES STREET  0.000000           0.333333             0.0   
125        SHELDON AVENUE  0.000000           0.000000             0.0   
126         SOUTH END ROW  0.000000           0.000000             0.0   
127   SOUTHWOOD LAWN ROAD  0.000000           0.000000             0.0   
128        SOVEREIGN PARK  0.000000           0.000000             0.0   
129      ST OSWALDS PLACE  0.000000           0.000000             0.0   
130      ST PETERS SQUARE  0.000000           0.000000             0.0   
131      STAFFORD TERRACE  0.000000           0.000000             0.0   
132      SUTHERLAND PLACE  0.000000           0.000000             0.0   
133         SYDNEY STREET  0.000000           0.000000             0.0   
134           THAMES BANK  0.000000           0.000000             0.0   
135           THE HEXAGON  0.000000           0.000000             0.0   
136       TREDEGAR SQUARE  0.000000           0.000000             0.0   
137        TRINITY STREET  0.000000           0.000000             0.0   
138  UPPER HAMPSTEAD WALK  0.000000           0.000000             0.0   
139       WALPOLE GARDENS  0.000000           0.000000             0.0   
140        WARWICK SQUARE  0.000000           0.000000             0.0   
141           WELBECK W

In [25]:
london_grouped.shape

(149, 349)

In [26]:
# Define a function to return the most common venues/facilities nearby real estate investments#
num_top_venues = 10

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [27]:
indicators = ['st', 'nd', 'rd']
num_top_venues = 10
# creaye columns accprding to number of top venues
columns = ['Street']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [28]:
# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Street'] = london_grouped['Street']

for ind in np.arange(london_grouped.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(london_grouped.iloc[ind, :], num_top_venues)

In [29]:
venues_sorted.head()

Street 1st Most Common Venue 2nd Most Common Venue  \
0    ALBION SQUARE                  Café     Indian Restaurant   
1      ANHALT ROAD                   Pub     French Restaurant   
2  ANSDELL TERRACE             Juice Bar                 Hotel   
3  APPLEGARTH ROAD                   Pub             Nightclub   
4  BARONSMEAD ROAD                   Pub       Thai Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub           Coffee Shop            Restaurant   
1         Grocery Store                 Diner                 Plaza   
2                   Pub            Restaurant     Indian Restaurant   
3                   Bar                Casino     Food & Drink Shop   
4           Pizza Place      Community Center            Restaurant   

  6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
0                   Bar          Burger Joint  New American Restaurant   
1   Japanese Restaurant  Gym / Fitness Center             Cocktail Bar   
2        Clothing Store    Italian Restaurant                     Café   
3                  Food          Food Service                  Exhibit   
4           Coffee Shop                  Park           Farmers Market   

  9th Most Common Venue    10th Most Common Venue  
0    Seafood Restaurant                    Museum  
1    English Restaurant                    Garden  
2           Supermarket  Mediterranean Restaurant  
3               Factory        Falafel Restaurant  
4                  Café           Nature Preserve

In [30]:
venues_sorted.shape

(149, 11)

In [31]:
london_grouped.shape

(149, 349)

In [32]:
london_grouped=df

After our inspection of venues, facilities and amenities nearby the most profitable real estate in Loondon, we cab now start clustering properties accordingly.

We will be distributing the area into 10 clusters.

In [46]:
# set number of clusters
kclusters = 10

london_grouped_clustering = london_grouped.drop('Street', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(london_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:50]

array([7, 1, 8, 9, 3, 4, 5, 2, 2, 7, 9, 9, 7, 4, 3, 3, 7, 1, 2, 4, 6, 0,
       9, 4, 7, 8, 9, 0, 7, 8, 9, 3, 1, 5, 3, 6, 9, 1, 3, 2, 4, 5, 0, 2,
       6, 2, 2, 6, 8, 8], dtype=int32)

In [47]:
#Dataframe to include Clusters

london_grouped_clustering=df
london_grouped_clustering.head()

Street  Avg_Price   Latitude   Longitude  Cluster Labels
196    ALBION SQUARE  2450000.0 -41.273758  173.289393               1
390      ANHALT ROAD  2435000.0  51.480316   -0.166801               3
405  ANSDELL TERRACE  2250000.0  51.499890   -0.189103               0
422  APPLEGARTH ROAD  2400000.0  53.748654   -0.326670               3
855  BARONSMEAD ROAD  2375000.0  51.477315   -0.239457               2

In [48]:
london_grouped_clustering.shape

(162, 5)

In [49]:
london_grouped_clustering.dtypes

Street             object
Avg_Price         float64
Latitude          float64
Longitude         float64
Cluster Labels      int32
dtype: object

In [50]:
df.shape

(162, 5)

In [51]:
df.dtypes

Street             object
Avg_Price         float64
Latitude          float64
Longitude         float64
Cluster Labels      int32
dtype: object

In [52]:

# add clustering labels
london_grouped_clustering['Cluster Labels'] = kmeans.labels_

# merge london_grouped with london_data to add latitude/longitude for each neighborhood
london_grouped_clustering = london_grouped_clustering.join(venues_sorted.set_index('Street'), on='Street')

london_grouped_clustering.head(30) # check the last columns!

Street     Avg_Price   Latitude   Longitude  \
196         ALBION SQUARE  2.450000e+06 -41.273758  173.289393   
390           ANHALT ROAD  2.435000e+06  51.480316   -0.166801   
405       ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103   
422       APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670   
855       BARONSMEAD ROAD  2.375000e+06  51.477315   -0.239457   
981        BEAUCLERC ROAD  2.480000e+06  30.211452  -81.617981   
1102      BELVEDERE DRIVE  2.340000e+06  44.707562  -63.545599   
1215    BICKENHALL STREET  2.208500e+06  51.521201   -0.158908   
1253    BIRCHLANDS AVENUE  2.217000e+06  51.448394   -0.160468   
1553       BRAMPTON GROVE  2.456875e+06  51.589961   -0.318525   
1632    BRIARDALE GARDENS  2.397132e+06  51.560175   -0.195431   
1797             BROOKWAY  2.400000e+06  45.432185 -122.802812   
1914         BURBAGE ROAD  2.445000e+06  52.538507   -1.353674   
1980            BURY WALK  2.492500e+06  52.145529   -0.423593   
2068      CALLCOTT STREET  2.375000e+06  51.508350   -0.198328   
2129    CAMPDEN HILL ROAD  2.379653e+06  51.506352   -0.198853   
2136         CAMPION ROAD  2.461000e+06  43.623321  -70.236225   
2158        CANNING PLACE  2.425000e+06  51.499494   -0.185481   
2225        CARLISLE ROAD  2.200000e+06  43.399438  -79.973215   
2230      CARLTON GARDENS  2.483500e+06 -37.801943  144.971970   
2242        CARLYLE COURT  2.300000e+06  32.972701  -97.173392   
2406       CHALCOT SQUARE  2.286679e+06  51.541088   -0.155928   
2484         CHARLES LANE  2.414000e+06  40.257626  -75.093891   
2562     CHELSEA CRESCENT  2.495000e+06  34.522443  -85.443891   
2606  CHESTER CLOSE NORTH  2.450000e+06  51.529205   -0.145081   
2638         CHEYNE COURT  2.250000e+06  52.035905    0.727612   
2641           CHEYNE ROW  2.410000e+06  51.483717   -0.169603   
2686        CHISWICK MALL  2.287500e+06  51.487994   -0.246605   
2761            CITY ROAD  2.468340e+06  51.529697   -0.097763   
2808     CLARENDON STREET  2.250000e+06  42.284961  -71.345534   

      Cluster Labels       1st Most Common Venue 2nd Most Common Venue  \
196                7                        Café     Indian Restaurant   
390                1                         Pub     French Restaurant   
405                8                   Juice Bar                 Hotel   
422                9                         Pub             Nightclub   
855                3                         Pub       Thai Restaurant   
981                4                         Spa       Automotive Shop   
1102               5                  Campground           Zoo Exhibit   
1215               2                   Gastropub            Restaurant   
1253               2                         Pub        Breakfast Spot   
1553               7                Food Service          Home Service   
1632               9           Indian Restaurant    Chinese Restaurant   
1797               9                         NaN                   NaN   
1914               7  Construction & Landscaping                   Bar   
1980               4                 Supermarket    English Restaurant   
2068               3                         Pub                  Park   
2129               3                         Pub                Bakery   
2136               7                         Pub                 Trail   
2158               1                       Hotel    Italian Restaurant   
2225               2                         NaN                   NaN   
2230               4          Italian Restaurant          Dessert Shop   
2242               6                        Farm            Soup Place   
2406               0                        Café                   Pub   
2484               9                         Spa           Dry Cleaner   
2562               4                         NaN                   NaN   
2606               7                        Park                Garden   
2638               8              Cosmetics Shop

In [53]:

# Create Map

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_grouped_clustering['Latitude'], london_grouped_clustering['Longitude'], london_grouped_clustering['Street'], london_grouped_clustering['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [54]:

london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 0, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
2406  2.286679e+06                  Café                   Pub   
2686  2.287500e+06                   Pub            Art Museum   
4285  2.265000e+06  Gym / Fitness Center   American Restaurant   
5486  2.270000e+06                   Pub                  Café   
5952  2.262500e+06                 River           Zoo Exhibit   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2406                   Bar           Coffee Shop    Italian Restaurant   
2686             Gift Shop  Gym / Fitness Center               Brewery   
4285                 Trail                   Gym  Fast Food Restaurant   
5486          Burger Joint            Restaurant                   Bar   
5952   Filipino Restaurant  Ethiopian Restaurant           Event Space   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2406     French Restaurant                  Park     Convenience Store   
2686          Cupcake Shop           Event Space               Factory   
4285    English Restaurant  Ethiopian Restaurant           Event Space   
5486          Cocktail Bar           Coffee Shop  Gym / Fitness Center   
5952               Exhibit               Factory    Falafel Restaurant   

     9th Most Common Venue 10th Most Common Venue  
2406                   Spa              Speakeasy  
2686       Coworking Space     Falafel Restaurant  
4285               Exhibit                Factory  
5486         Grocery Store     Italian Restaurant  
5952                  Farm         Farmers Market

In [55]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 1, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
390   2.435000e+06                   Pub     French Restaurant   
2158  2.425000e+06                 Hotel    Italian Restaurant   
2995  2.440000e+06     Indian Restaurant      Department Store   
3848  2.425000e+06                   Pub    Weight Loss Center   
5289  2.441667e+06                   Pub                 Hotel   

     3rd Most Common Venue 4th Most Common Venue  \
390          Grocery Store                 Diner   
2158             Juice Bar     Indian Restaurant   
2995         Shopping Mall        Shipping Store   
3848         Grocery Store     Food & Drink Shop   
5289     Convention Center                  Park   

              5th Most Common Venue     6th Most Common Venue  \
390                           Plaza       Japanese Restaurant   
2158             Chinese Restaurant  Mediterranean Restaurant   
2995  Paper / Office Supplies Store              Liquor Store   
3848                           Food        English Restaurant   
5289                 Sandwich Place                 Gastropub   

     7th Most Common Venue 8th Most Common Venue     9th Most Common Venue  \
390   Gym / Fitness Center          Cocktail Bar        English Restaurant   
2158     French Restaurant            Restaurant            Clothing Store   
2995            Laundromat              Pharmacy               Coffee Shop   
3848  Ethiopian Restaurant           Event Space                   Exhibit   
5289    Italian Restaurant     Indian Restaurant  Bike Rental / Bike Share   

     10th Most Common Venue  
390                  Garden  
2158                    Pub  
2995                 Resort  
3848                Factory  
5289                    Bar

In [56]:

london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 2, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
1215  2208500.0             Gastropub            Restaurant   
1253  2217000.0                   Pub        Breakfast Spot   
2225  2200000.0                   NaN                   NaN   
3980  2225000.0                  Café                   Pub   
4289  2200000.0     Indian Restaurant         Grocery Store   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1215           Coffee Shop           Pizza Place                 Hotel   
1253           Coffee Shop               Brewery     French Restaurant   
2225                   NaN                   NaN                   NaN   
3980    Italian Restaurant                Bakery           Coffee Shop   
4289        Sandwich Place                  Café       Thai Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1215    Italian Restaurant                Garden                Bakery   
1253    Chinese Restaurant                  Lake         Train Station   
2225                   NaN                   NaN                   NaN   
3980                Museum       Thai Restaurant             Bookstore   
4289                   Bar                   Pub          Home Service   

            9th Most Common Venue 10th Most Common Venue  
1215             Greek Restaurant                    Bar  
1253                       Bakery            Pizza Place  
2225                          NaN                    NaN  
3980                       Bistro          Grocery Store  
4289  Eastern European Restaurant      Electronics Store

In [57]:

london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 3, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
855   2375000.0                   Pub       Thai Restaurant   
2068  2375000.0                   Pub                  Park   
2129  2379652.7                   Pub                Bakery   
2944  2367500.0                 Hotel                   Pub   
3255  2360000.0                   Pub         Grocery Store   

     3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
855            Pizza Place      Community Center              Restaurant   
2068        Ice Cream Shop                Bakery       Indian Restaurant   
2129     Indian Restaurant        Ice Cream Shop             Coffee Shop   
2944                Garden                  Café             Coffee Shop   
3255           Pizza Place                  Café  Furniture / Home Store   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
855            Coffee Shop                  Park        Farmers Market   
2068                 Hotel           Yoga Studio          Burger Joint   
2129                  Park                 Hotel                Hostel   
2944    Italian Restaurant    Chinese Restaurant                   Bar   
3255      Greek Restaurant           Art Gallery    Italian Restaurant   

         9th Most Common Venue 10th Most Common Venue  
855                       Café        Nature Preserve  
2068      Caribbean Restaurant         Sandwich Place  
2129               Yoga Studio         Breakfast Spot  
2944  Mediterranean Restaurant             Restaurant  
3255                    Bakery              Wine Shop

In [58]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 4, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
981   2480000.0                   Spa       Automotive Shop   
1980  2492500.0           Supermarket    English Restaurant   
2230  2483500.0    Italian Restaurant          Dessert Shop   
2562  2495000.0                   NaN                   NaN   
4035  2477500.0           Coffee Shop                   Pub   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
981        Harbor / Marina           Pizza Place           Zoo Exhibit   
1980   Rental Car Location                   Gym        Hardware Store   
2230   Lebanese Restaurant     Convenience Store   Japanese Restaurant   
2562                   NaN                   NaN                   NaN   
4035                   Bar          Cocktail Bar    Italian Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
981    Filipino Restaurant           Event Space               Exhibit   
1980  Fast Food Restaurant           Coffee Shop                  Café   
2230           Coffee Shop    Light Rail Station        Ice Cream Shop   
2562                   NaN                   NaN                   NaN   
4035       Thai Restaurant                 Hotel     French Restaurant   

     9th Most Common Venue 10th Most Common Venue  
981                Factory     Falafel Restaurant  
1980           Dry Cleaner                   Park  
2230         Deli / Bodega                  Hotel  
2562                   NaN                    NaN  
4035                Museum     English Restaurant

In [59]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 5, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
1102  2.340000e+06            Campground           Zoo Exhibit   
3202  2.350000e+06                  Café                Museum   
4232  2.335500e+06                 Hotel                   Pub   
5382  2.350833e+06                   NaN                   NaN   
5493  2.325000e+06     Indian Restaurant                  Café   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1102           Film Studio               Exhibit               Factory   
3202           Coffee Shop                Garden                  Park   
4232        Sandwich Place                  Café    Italian Restaurant   
5382                   NaN                   NaN                   NaN   
5493           Film Studio               Exhibit               Factory   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1102    Falafel Restaurant                  Farm        Farmers Market   
3202       Thai Restaurant           Pizza Place          Tennis Court   
4232           Coffee Shop                Bakery            Restaurant   
5382                   NaN                   NaN                   NaN   
5493    Falafel Restaurant                  Farm        Farmers Market   

     9th Most Common Venue 10th Most Common Venue  
1102  Fast Food Restaurant    Filipino Restaurant  
3202         Movie Theater                   Lake  
4232  Gym / Fitness Center     Turkish Restaurant  
5382                   NaN                    NaN  
5493  Fast Food Restaurant    Filipino Restaurant

In [60]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 6, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
2242  2300000.0                  Farm            Soup Place   
3377  2298000.0                 Hotel           Zoo Exhibit   
4375  2320425.0                  Café           Supermarket   
5098  2299000.0                 Hotel                   Pub   
6227  2310000.0                   Pub                  Café   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2242           Zoo Exhibit           Film Studio           Event Space   
3377           Film Studio           Event Space               Exhibit   
4375              Platform                  Park  Gym / Fitness Center   
5098           Coffee Shop    Chinese Restaurant                Garden   
6227  Gym / Fitness Center     Indian Restaurant    Persian Restaurant   

     6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
2242               Exhibit               Factory      Falafel Restaurant   
3377               Factory    Falafel Restaurant                    Farm   
4375      Stationery Store           Bus Station  Furniture / Home Store   
5098                  Café      Malay Restaurant       Indian Restaurant   
6227                Garden           Pizza Place              Restaurant   

     9th Most Common Venue 10th Most Common Venue  
2242        Farmers Market   Fast Food Restaurant  
3377        Farmers Market   Fast Food Restaurant  
4375  Fast Food Restaurant         Breakfast Spot  
5098           Pizza Place            Supermarket  
6227           Coffee Shop                  Hotel

In [61]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 7, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price       1st Most Common Venue 2nd Most Common Venue  \
196   2450000.0                        Café     Indian Restaurant   
1553  2456875.0                Food Service          Home Service   
1914  2445000.0  Construction & Landscaping                   Bar   
2136  2461000.0                         Pub                 Trail   
2606  2450000.0                        Park                Garden   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
196                    Pub           Coffee Shop            Restaurant   
1553           Zoo Exhibit           Film Studio           Event Space   
1914         Grocery Store          Dance Studio    Athletics & Sports   
2136                  Food  Fast Food Restaurant  Ethiopian Restaurant   
2606          Cocktail Bar              Beer Bar            Playground   

     6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
196                    Bar          Burger Joint  New American Restaurant   
1553               Exhibit               Factory       Falafel Restaurant   
1914           Film Studio               Exhibit                  Factory   
2136           Event Space               Exhibit                  Factory   
2606  Gym / Fitness Center           Coffee Shop       Mexican Restaurant   

       9th Most Common Venue 10th Most Common Venue  
196       Seafood Restaurant                 Museum  
1553                    Farm         Farmers Market  
1914      Falafel Restaurant                   Farm  
2136      Falafel Restaurant                   Farm  
2606  Furniture / Home Store  Performing Arts Venue

In [62]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 8, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue       2nd Most Common Venue  \
405   2250000.0             Juice Bar                       Hotel   
2638  2250000.0        Cosmetics Shop          Italian Restaurant   
2808  2250000.0           Pizza Place          Italian Restaurant   
5241  2258500.0      Business Service           Fish & Chips Shop   
5244  2242500.0     Indian Restaurant  Construction & Landscaping   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
405                    Pub            Restaurant     Indian Restaurant   
2638             Bookstore           Supermarket             Gift Shop   
2808          Liquor Store                  Park    Seafood Restaurant   
5241     Convenience Store               Exhibit               Factory   
5244    Photography Studio           Pizza Place              Bus Stop   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
405          Clothing Store    Italian Restaurant                  Café   
2638                 Bakery              Pharmacy        Clothing Store   
2808  Performing Arts Venue                  Bank    Chinese Restaurant   
5241     Falafel Restaurant                  Farm        Farmers Market   
5244                   Food           Flower Shop               Exhibit   

     9th Most Common Venue    10th Most Common Venue  
405            Supermarket  Mediterranean Restaurant  
2638           Coffee Shop             Grocery Store  
2808        Sandwich Place            Farmers Market  
5241  Fast Food Restaurant       Filipino Restaurant  
5244               Factory        Falafel Restaurant

In [63]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 9, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
422   2400000.0                   Pub             Nightclub   
1632  2397132.0     Indian Restaurant    Chinese Restaurant   
1797  2400000.0                   NaN                   NaN   
2484  2414000.0                   Spa           Dry Cleaner   
2641  2410000.0                  Café                   Pub   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
422                       Bar                Casino     Food & Drink Shop   
1632  Health & Beauty Service           Coffee Shop         Grocery Store   
1797                      NaN                   NaN                   NaN   
2484      American Restaurant    Photography Studio           Zoo Exhibit   
2641       Italian Restaurant             Nightclub                Bakery   

     6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
422                   Food          Food Service                 Exhibit   
1632  Fast Food Restaurant  Ethiopian Restaurant             Event Space   
1797                   NaN                   NaN                     NaN   
2484   Filipino Restaurant               Exhibit                 Factory   
2641             Juice Bar  Gym / Fitness Center  Furniture / Home Store   

     9th Most Common Venue 10th Most Common Venue  
422                Factory     Falafel Restaurant  
1632               Exhibit                Factory  
1797                   NaN                    NaN  
2484    Falafel Restaurant                   Farm  
2641           Pizza Place                  Plaza

## Results and Discussion Section

We may analyze our results according to the clusterd that we produced. Even though, all the clusters could proise an optimal range of facilities and amenities, we have foun certainn patterns.

Clusters 0, 1 and 3 are for the ones who enjoy spendind weekends in clubs and pubs, while cluster 5 and 6 are for theatre lovers.
Investing in clusters 4, 8 and 9 would be the best as they provide the maximum facilities in their neighbourhoods. For the home buyers who are prone to live in 'green' areas should go for areas in cluster 2, 5, 6 and 7 as they have parks, gardens, farms and waterfronts.



## Conclusion

To sum up, in a country with such a huge area and high real estate prices, keeping in mind all the amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores, we aimed at finding an appropriate location to incest in. As a result, we came up with a project, where we recomment the investers a perfect location.

To solve the business problem, we clustered London neighborhoods in order to recommend areas in the required budget. We recommended profitable venues according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores.

First, we gathered data on London properties and the relative price paid data were extracted from the HM Land Registry (http://landregistry.data.gov.uk/). We accessed data through FourSquare API interface and arranged them as a data frame for visualization. By merging data on London properties and the relative price paid data from the HM Land Registry and data on amenities and essential facilities surrounding such properties from FourSquare API interface, we were able to recommend profitable real estate investments.

Second, The Methodology section comprised four stages: 1. Collect Inspection Data; 2. Explore and Understand Data; 3. Data preparation and preprocessing; 4. Modeling. In particular, in the modeling section, we used the k-means clustering technique as it is fast and efficient in terms of computational cost, is highly flexible to account for mutations in real estate market in London and is accurate.

Finally, we concluded recommending the locations to investors according to their interests.